src/structure.js

```diff
@@ -66,21 +66,65 @@ function Structure (options) {
       Infinity;
   }
 
-  this._pathListeners = Immutable.Map();
-  this.on('swap', function (newData, oldData, keyPath) {
-    var listenerData = listMatchingOrCreateEmptyList(self._pathListeners, keyPath);
-    self._pathListeners = listenerData.listeners;
+   this._referencelisteners = Immutable.Map();
 
-    listenerData.matched.forEach(function (fn) {
-      fn(keyPath, newData, oldData);
-    });
+  this.on('swap', function (newData, oldData, keyPath) {
+      var args = [keyPath, newData, oldData];
+      if (!keyPath || keyPath.length === 0) {
+          emit(
+              self._referencelisteners,
+              newData,
+              oldData,
+              [],
+              args
+          );
+      } else {
+          var path = keyPath[0];
+          emit(
+              self._referencelisteners.get(path),
+              newData.get(path),
+              oldData.get(path),
+              keyPath.slice(1),
+              args
+          );
+      }
   });
 
   EventEmitter.call(this, arguments);
 }
 inherits(Structure, EventEmitter);
 module.exports = Structure;
+var FUNCTION_KEY = {};
+
+function emit(map, newData, oldData, path, args) {
+
+    if (map && newData !== oldData) {
+        map.get(FUNCTION_KEY, []).forEach(function (fn) {
+            fn.apply(null, args);
```

Comment:
We should probably pass the "calculated" `keyPath`?

> This may lead to issues if the observing function is pure, and expect only changes at "someBox" path.

But it will only trigger when `someBox` or any of `someBox`s children has changed? And if children of `someBox` has changed, `someBox` has changed, no?